### PR TITLE
perf(blob): Skip redundant Touch of healthy blobs on HEAD/PUT probes

### PR DESCRIPTION
**Pull request by Vad1mo**
### What this PR does

Skips the `Touch()` write for blobs that are already `StatusNone` when their `update_time` is fresher than half the GC time window. The `StatusDelete`/`StatusDeleteFailed` → `StatusNone` transitions (rescuing a blob from GC candidacy) are unchanged and always written.

### Why

Every `HEAD /v2/<name>/blobs/<digest>` and every blob/manifest PUT probe calls `Touch()` on the blob row, even when the blob is already healthy. For a `StatusNone` blob the status transition is a no-op — the write exists only to refresh `update_time` so the GC useless-blob sweep (`update_time <= now() - GC time window`) cannot select a blob that is still in use.

Under concurrent pushes of images sharing base layers, every client probes the same few blob rows and each probe rewrites them:

```
UPDATE blob SET version = version + 1, update_time = $1, status = $2 WHERE id = $3 AND version >= $4 AND status IN (...)
```

Because the `StatusNone` case uses `version >= ?`, the write always succeeds — it never conflicts, it just serializes (row lock queueing). Measured on a production Harbor (v2.15.x) during a ~1000-blob CI push burst, this statement accounted for **6.9 average active sessions** on a 2-vCPU database, concentrated on a handful of shared base-layer rows.

### Correctness, including the GC race in goharbor/harbor#23116

The GC data-loss race described in goharbor/harbor#23116 involves a blob whose `update_time` is **stale** (older than the GC window) while it is being reused. This change does not affect that case: a stale timestamp always fails the freshness check, so `Touch` still fires exactly as today — the rescue write is preserved precisely where it matters. Only writes to rows touched within the last half-window are skipped, and such rows are by construction at least half a window away from GC's selection boundary, so GC can never select a blob that received a skipped probe.

With the default 2h window: a healthy blob is re-touched at most 1 hour after its last write, leaving a full hour of margin. The freshness comparison mirrors the one the adjacent `StatusDeleting` branch already performs against `GetGCTimeWindow()`.

A boundary unit test is included (`TestTouchIsDue`), covering the just-under/just-over threshold cases and the stale case.

### Related Issues

- Related to #404 — the remaining rescue/keepalive `Touch` still runs inside the request-wide transaction; that follow-up stays open after this PR.
- Overlaps with #397, which implements the same half-window debounce for the same probe paths — the two should be reconciled before merging.

### CODE REVIEW

File: src/server/middleware/blob/head_blob.go [hunk 1]
```diff
@@ -55,7 +55,19 @@ func handleHead(req *http.Request) error {
 	}
 
 	switch bb.Status {
-	case blob_models.StatusNone, blob_models.StatusDelete:
+	case blob_models.StatusNone:
+		// The blob is already healthy; Touch only refreshes update_time to
+		// keep it out of GC's candidate window, so skip the write while the
+		// timestamp is still fresh (see touchIsDue).
+		if touchIsDue(bb) {
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/head_blob.go]: The skip only ever happens while update_time is younger than half the GC window, so a skipped HEAD always leaves at least W/2 (1h with the 2h default) before the blob can enter GC's candidate window. The previous code gave a full W from every probe, so the worst-case margin is halved, not removed - that is the deliberate trade for eliminating the hottest write on the burst path. A client that waits more than an hour between the HEAD and the referencing manifest PUT was already racing GC before this change (GC must also actually run in that window and the blob must be unreferenced in every project). Restoring the full-window guarantee is better done by making the GC candidate query aware of references than by reintroducing the per-probe row write.

File: src/server/middleware/blob/util.go [hunk 2]
```diff
@@ -40,7 +40,19 @@ func probeBlob(r *http.Request, digest string) error {
 	}
 
 	switch bb.Status {
-	case models.StatusNone, models.StatusDelete, models.StatusDeleteFailed:
+	case models.StatusNone:
+		// The blob is already healthy; Touch only refreshes update_time to
+		// keep it out of GC's candidate window, so skip the write while the
+		// timestamp is still fresh (see touchIsDue).
+		if touchIsDue(bb) {
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/util.go]: Same margin argument as the HEAD path (see the reply on head_blob.go): a skipped probe implies update_time is younger than W/2, so at least an hour (default window) remains before GC eligibility. An upload that takes longer than that between probe and association was already racing the 2h window before this change. Blobs in StatusDelete/StatusDeleteFailed - the case where a stale timestamp is actually dangerous - are always touched.

File: src/server/middleware/blob/util.go [hunk 3]
```diff
@@ -62,3 +74,17 @@ func probeBlob(r *http.Request, digest string) error {
...
+// stale enough that Touch must rewrite it. For StatusNone blobs the status
+// transition is a no-op: the write exists only to refresh update_time so
+// that the GC useless-blob sweep (update_time <= now() - GC time window)
+// cannot select a blob that is still in use. If the row was written within
+// the last half window, the stored timestamp is still at least half a
+// window away from the GC boundary and rewriting it changes nothing.
+// Under concurrent pushes of images sharing base layers, this redundant
+// write is the hottest row contention on the database (every HEAD and PUT
+// of every shared layer), so it is skipped while fresh.
+func touchIsDue(bb *models.Blob) bool {
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/util.go:82]: Correct description of the trade: worst-case margin goes from W to W/2 (1h by default). That is intentional - see the replies on the two probe sites. The freshly-created-blob case is not affected: a brand-new blob row was just written, so its update_time is now and the full window applies; the halved margin only concerns blobs whose last write was already up to W/2 ago, which the old code would re-touch on every probe at the cost of the hottest row contention in the incident.

File: src/server/middleware/blob/util_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+func TestTouchIsDue(t *testing.T) {
+	// default GC time window is 2 hours, so the debounce threshold is 1 hour
+	cases := []struct {
+		name string
+		age  time.Duration
+		due  bool
+	}{
+		{"just written", 10 * time.Second, false},
+		{"well within threshold", 30 * time.Minute, false},
+		{"just under threshold", 59 * time.Minute, false},
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/util_test.go]: Fixed in 571f898 - the cases are now derived from config.GetGCTimeWindow() instead of hardcoding the 2h default.

File: src/server/middleware/blob/util_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		{"just over threshold", 61 * time.Minute, true},
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/util_test.go]: Fixed in 571f898 together with the sibling thread - thresholds are computed from the configured window.

File: src/server/middleware/blob/util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/util.go]: Extracted a shared touchIfDue helper used by both middleware paths in 571f898.

File: src/server/middleware/blob/util_test.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/goharbor/harbor/src/lib/config"
+	"github.com/goharbor/harbor/src/pkg/blob/models"
+)
+
+func TestTouchIsDue(t *testing.T) {
+	// derive the cases from the configured window (2h by default, but
+	// GC_TIME_WINDOW_HOURS may override it in the environment) so the
+	// expectations track exactly what touchIsDue reads at runtime
+	window := time.Duration(config.GetGCTimeWindow()) * time.Hour
+	threshold := window / 2
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/server/middleware/blob/util_test.go:38]: Fixed in 97b1add - the test now skips when the configured window is zero, since GC_TIME_WINDOW_HOURS=0 disables the debounce entirely and the boundary cases only exist for a positive window.

**Comment by bupd** [comment]
Superseded by #397 (merged as 2004d1981d), which implements the same half-window Touch debounce with DB-backed middleware tests. The stack has been rebased so #718–#720 no longer include this commit; the only unique bit here (the shared touchIfDue helper extraction) can be revisited as a small refactor later if wanted.
